File: automata/automata_converter.py
from collections import deque
from .state import State
from .state_type import StateType
from .transition import Transition
from .nfa import NFA
from .dfa import DFA


class AutomataConverter:
    def __init__(self, nfa):
        self.nfa = nfa
        self.transition_table = []
        self.handled_states = set()
        self.unhandled_states = set()
        self.phi_state_exist = False
        self.PHI = 'Ø'

    # def convert_to_dfa(self):
    #     if self.nfa.has_epsilon():
    #         self.nfa = self.convert_to_nfa(self.nfa)
    #     self.add_table_row(self.nfa.get_initial_state())
    #     self.process_unhandled_states()
    #
    #     if self.phi_state_exist:
    #         self.add_phi_row()
    #
    #     return self.transition_table

    def convert_to_dfa(self):
        self.handled_states.clear()
        dfa = DFA()
        dfa.set_alphabets(self.nfa.get_alphabets())
        if (dfa.get_alphabets().__contains__('e')):
            dfa.alphabets.remove('e')

        phi = State(self.PHI)
        if self.nfa.has_epsilon:
            self.nfa = self.convert_to_nfa(self.nfa)

        start_state = self.nfa.get_initial_state()
        new_state = State(start_state.name)
        dfa.add_state(new_state)
        dfa.set_initial_state(new_state)

        if start_state in self.nfa.get_final_states():
            dfa.add_final_state(new_state)

        self.handled_states.add(start_state)

        for alphabet in self.nfa.get_alphabets():
            next_states = self.get_next_states_for_alphabet(start_state, alphabet)
            if next_states:
                if len(next_states) == 1:
                    new_state.add_transition(Transition(alphabet, next_states[0]))
                    if not self.is_state_exist(self.handled_states, next_states[0]):
                        self.unhandled_states.add(next_states[0])
                else:
                    combined_state = self.create_combined_state(next_states)
                    new_state.add_transition(Transition(alphabet, combined_state))
            else:
                self.phi_state_exist = True
                new_state.add_transition(Transition(alphabet, phi))

        while self.unhandled_states:
            state = self.unhandled_states.pop()
            new_state = State(state.name)

            for alphabet in self.nfa.get_alphabets():
                next_states = self.get_next_states_for_alphabet(state, alphabet)
                if next_states:
                    if len(next_states) == 1:
                        new_state.add_transition(Transition(alphabet, next_states[0]))
                        if not self.is_state_exist(self.handled_states, next_states[0]):
                            self.unhandled_states.add(next_states[0])
                    else:
                        combined_state = self.create_combined_state(next_states)
                        new_state.add_transition(Transition(alphabet, combined_state))
                else:
                    self.phi_state_exist = True
                    new_state.add_transition(Transition(alphabet, phi))

            self.handled_states.add(state)
            dfa.add_state(new_state)

            if state in self.nfa.get_final_states():
                dfa.add_final_state(new_state)

        if self.phi_state_exist:
            dfa.add_state(phi)
            for alphabet in self.nfa.get_alphabets():
                dfa.add_transition(phi, alphabet, phi)

        return dfa

    def add_table_row(self, state):
        row = [state]
        self.handled_states.add(state)

        for symbol in self.nfa.get_alphabets():
            next_states = self.get_next_states_for_alphabet(state, symbol)
            if next_states:
                if len(next_states) == 1:
                    new_state = next_states[0]
                    if not self.is_state_exist(self.handled_states, new_state):
                        self.unhandled_states.add(new_state)
                    row.append(new_state)
                else:
                    new_state = self.create_combined_state(next_states)
                    row.append(new_state)
            else:
                phi = State(self.PHI)
                self.phi_state_exist = True
                row.append(phi)

        self.transition_table.append(row)

    def get_next_states_for_alphabet(self, state, alphabet):
        return [transition.get_next_state() for transition in state.get_transitions() if transition.get_alphabet() == alphabet]

    def create_combined_state(self, next_states):
        sorted_names = sorted(state.name for state in next_states)
        new_state_name = "".join(sorted_names)

        for state in self.handled_states | self.unhandled_states:
            if state.name == new_state_name:
                return state

        new_state = State(new_state_name)
        if any(self.nfa.get_state_type(state) == StateType.FINAL for state in next_states):
            self.nfa.add_final_state(new_state)

        for state in next_states:
            for transition in state.get_transitions():
                new_state.add_transition(transition)

        self.unhandled_states.add(new_state)
        return new_state

    def process_unhandled_states(self):
        while self.unhandled_states:
            state = self.unhandled_states.pop()
            self.add_table_row(state)

    def convert_to_nfa(self, epsilon_nfa):
        converted_nfa = NFA()
        initial_state = State(epsilon_nfa.get_initial_state().name)
        converted_nfa.add_state(initial_state)
        converted_nfa.set_initial_state(initial_state)

        for curr in epsilon_nfa.get_states():
            current_state = converted_nfa.get_state_by_name(curr.name)
            if current_state:
                for symbol in epsilon_nfa.get_alphabets():
                    epsilon = set()
                    next_states = []

                    for state in self.get_epsilon_closure(curr):
                        next_states.extend(self.get_next_states_for_alphabet(state, symbol))

                    for next_state in next_states:
                        epsilon.update(self.get_epsilon_closure(next_state))

                    for state in epsilon:
                        new_state = converted_nfa.get_state_by_name(state.name)
                        if symbol != 'e':
                            converted_nfa.add_transition(current_state, symbol, new_state)
                            if epsilon_nfa.get_state_type(state) == StateType.FINAL:
                                converted_nfa.add_final_state(new_state)

        return converted_nfa

    def get_epsilon_closure(self, state):
        epsilon_closure = set()
        visited = set()
        stack = deque([state])

        while stack:
            current_state = stack.pop()
            if current_state not in visited:
                visited.add(current_state)
                epsilon_closure.add(current_state)

                for transition in current_state.get_transitions():
                    if transition.get_alphabet() == 'e':
                        stack.append(transition.get_next_state())

        return epsilon_closure

    def is_state_exist(self, states, state):
        return any(s.name == state.name for s in states)

    def add_phi_row(self):
        phi = State(self.PHI)
        row = [phi] * (len(self.nfa.get_alphabets()) + 1)
        self.transition_table.append(row)

File: src/AutomataConverter.java
import java.util.*;

public class AutomataConverter {
    private NFA nfa;
    private List<List<State>> transitionTable;
    private Set<State> handledStates;
    private Set<State> unhandledStates;
    private boolean phiStateExist;
    private boolean minimized;
    private String PHI = "?";


    public AutomataConverter(NFA nfa, boolean minimized) {
        this.nfa = nfa;
        transitionTable = new ArrayList<>();
        handledStates = new HashSet<>();
        unhandledStates = new LinkedHashSet<>();
        phiStateExist = false;
        this.minimized = minimized;
    }

    public List<List<State>> convertToDFA() {
        if (nfa.hasEpsilon()){
            nfa = convertToNFA(nfa);
//            System.out.println(nfa);
        }
        addTableRow(nfa.getInitialState());

        processUnhandledStates();
//        getRemainingStates();

        if (phiStateExist)
            addPhiRow();
        return transitionTable;
    }

    public DFA convertToDFA(boolean minimized) {
        handledStates.clear();
        DFA dfa = new DFA();
        dfa.setAlphabets(nfa.getAlphabets());
        if (dfa.getAlphabets().contains('e'))
            dfa.getAlphabets().remove('e');

        State phi = new State(PHI);
        if (nfa.hasEpsilon()){
            nfa = convertToNFA(nfa);
//            System.out.println(nfa);
        }
        State startState = nfa.getInitialState();
        State newS = new State(startState.getName());
        dfa.addState(newS);
        dfa.setInitialState(newS);
        if (nfa.getFinalStates().contains(startState))
            dfa.addFinalState(newS);
        handledStates.add(startState);

        for (char alphabet : nfa.getAlphabets()) {
            List<State> nextStates = getNextStatesForAlphabet(startState, alphabet);
            if (!nextStates.isEmpty()) {
                State newState;
                if (nextStates.size() == 1) {
                    newState = nextStates.get(0);
                    if (!isStateExist(handledStates, newState))
                        unhandledStates.add(newState);
                    newS.addTransition(new Transition(alphabet, nextStates.get(0)));
                } else {
//                    removeAllTransitionsForAlphabet(startState, alphabet);
                    newState = createCombinedState(nextStates);
                    newS.addTransition(new Transition(alphabet, newState));
                }
            } else {
                phiStateExist = true;
                newS.addTransition(new Transition(alphabet, phi));
            }
        }

        while (!unhandledStates.isEmpty()) {
            Iterator<State> iterator = unhandledStates.iterator();
            State state = iterator.next();
            State newState = new State(state.getName());

            for (char alphabet : nfa.getAlphabets()) {
                List<State> nextStates = getNextStatesForAlphabet(state, alphabet);
                if (!nextStates.isEmpty()) {

                    if (nextStates.size() == 1) {
                        newState.addTransition(new Transition(alphabet, nextStates.get(0)));
                        if (!isStateExist(handledStates, nextStates.get(0)))
                            unhandledStates.add(nextStates.get(0));

                    } else {
                        newState.addTransition(new Transition(alphabet, createCombinedState(nextStates)));
                    }
                } else {
                    phiStateExist = true;
                    newState.addTransition(new Transition(alphabet, phi));
                }
            }
            unhandledStates.remove(state);
            handledStates.add(state);
            dfa.addState(newState);
            if (nfa.getFinalStates().contains(state))
                dfa.addFinalState(newState);
        }

        if (phiStateExist) {
            dfa.addState(phi);
            for (char alpha: nfa.getAlphabets())
                dfa.addTransition(phi, alpha, phi);
        }
        return dfa;
    }

    private void addTableRow(State state) {
        List<State> row = new ArrayList<>();
        row.add(state);
        handledStates.add(state);

        for (char symbol : nfa.getAlphabets()) {
            List<State> nextStates = getNextStatesForAlphabet(state, symbol);
            if (!nextStates.isEmpty()) {
                State newState;
                if (nextStates.size() == 1) {
                    newState = nextStates.get(0);
                    if (!isStateExist(handledStates, newState))
                        unhandledStates.add(newState);
                    row.add(newState);
                } else {
                    newState = createCombinedState(nextStates);

                    row.add(newState);
                }
            } else {
                State phi = new State(PHI);
                phiStateExist = true;
                row.add(phi);
            }
        }

        transitionTable.add(row);
    }

    private List<State> getNextStatesForAlphabet(State state, char alphabet) {
        List<State> nextStates = new ArrayList<>();
        for (Transition transition : state.getTransitions()) {
            if (transition.getAlphabet() == alphabet) {
                nextStates.add(transition.getNextState());
//                if (!isStateExist(handledStates, transition.getNextState()))
//                    unhandledStates.add(state);
            }
        }
        return nextStates;
    }

    private void removeAllTransitionsForAlphabet(State state, char alphabet) {
        for (Transition transition : state.getTransitions()) {
            if (transition.getAlphabet() == alphabet) {
                state.getTransitions().remove(transition);
            }
        }}

    private State createCombinedState(List<State> nextStates) {
        /* using TreeSet because it sorts the new state alphabetically to make sure that
        there is no repeated state with another name
        */
        Set<String> sortedName = new TreeSet<>();
        StateType stateType = null;

        for (State nextState : nextStates){
            sortedName.add(nextState.getName());

            if (nfa.getStateType(nextState) == StateType.FINAL)
                stateType = StateType.FINAL;
        }
        String newStateName = String.join("", sortedName.toArray(new String[0]));

//        StringBuilder newStateName = new StringBuilder();

//        for (String nextState : sortedName)
//            newStateName.append(nextState).append("+");
//
//        newStateName.setLength(newStateName.length() - 1); // remove the last char '+'

//        for (State nextState : nextStates) {
//            if (minimized) {
//                if (!transitionTable.isEmpty()) // check if not the first row (solve a problem happened in slide 6)
//                    handledStates.add(nextState);
//            }
//        }


        // check if the new composited state is in handled or unhandled list so no need to create new one
        for (State s: handledStates){
            if (s.getName().equals(newStateName.toString()))
                return s;
        }
        for (State s: unhandledStates){
            if (s.getName().equals(newStateName.toString()))
                return s;
        }

        State newState = new State(newStateName);
        if (stateType == StateType.FINAL)
            nfa.addFinalState(newState);
        for (State nextState : nextStates) {
            for (Transition t : nextState.getTransitions()) {
                newState.addTransition(t);
            }
        }
        unhandledStates.add(newState);
        return newState;
    }

    private void getRemainingStates() {
        for(State state:nfa.getStates()){
            if (!isStateExist(handledStates, state))
                unhandledStates.add(state);
        }
        processUnhandledStates();
    }

    private NFA convertToNFA(NFA epsilonNFA) {
        NFA convertedNFA = new NFA();
        State initialState = new State(epsilonNFA.getInitialState().getName());
        convertedNFA.addState(initialState);
        convertedNFA.setInitialState(initialState);
        for(State curr: epsilonNFA.getStates()) {
            State currentState = convertedNFA.getStateByName(curr.getName());
            for (char symbol : epsilonNFA.getAlphabets()) {
                Set<State> epsilon = new HashSet<>();
                List<State> nextStates = new ArrayList<>();

                for (State state : getEpsilonClosure(curr)) {
                    for (State state2 : getNextStatesForAlphabet(state, symbol)) {
                        nextStates.add(state2);
                    }
                }
                for (State s : nextStates) {
                    for (State state : getEpsilonClosure(s)) {
                        epsilon.add(state);
                    }
                }
                for (State s : epsilon) {
                    State newState = convertedNFA.getStateByName(s.getName());
//                    convertedNFA.addState(newState);
                    convertedNFA.addTransition(currentState, symbol, newState);
                    if (epsilonNFA.getStateType(s) == StateType.FINAL)
                        convertedNFA.addFinalState(newState);
                }
            }
        }
        return convertedNFA;
    }

    private Set<State> getEpsilonClosure(State state){
        Set<State> epsilonClosure = new HashSet<>();
        Set<State> visited = new HashSet<>();
        Stack<State> stack = new Stack<>();
        stack.push(state);

        while (!stack.isEmpty()){
            State currentState = stack.pop();
            if (!visited.contains(currentState)){
                visited.add(currentState);
                epsilonClosure.add(currentState);

                for (Transition transition : currentState.getTransitions()) {
                    if (transition.getAlphabet() == 'e') { // check if there is any epsilon transition
//                        epsilonClosure.add(transition.getNextState());
                        stack.push(transition.getNextState());
                    }
                }
            }
        }


        return epsilonClosure;
    }

    private void processUnhandledStates() {
        while (!unhandledStates.isEmpty()) {
            Iterator<State> iterator = unhandledStates.iterator();
            State state = iterator.next();

//            if (!isStateExist(handledStates, state)) {
//                handledStates.add(state);
//                unhandledStates.remove(state);
//
//                addTableRow(state);
//            }

            addTableRow(state);
            unhandledStates.remove(state);
        }
//        getRemainingStates();
    }

    private boolean isStateExist(Set<State> states, State s){
        for (State state:states){
            if (state.getName().equals(s.getName()))
                return true;
        }
        return false;
    }

    private void addPhiRow () {
        State phi = new State(PHI);
        List<State> row = new ArrayList<>();
        for (int i = 0; i <= nfa.getAlphabets().size(); i++)
            row.add(phi);

        transitionTable.add(row);
    }

    public NFA getNfa() {
        return nfa;
    }
}